### ISSUE TITLE
(lambda-event-sources): support FunctionResponseTypes

### ISSUE THREAD

**Issue by mariobittencourt** [aws/aws-cdk#12654]
It is possible to enable ReportBatchItemFailures for DynamoDB when used as trigger for lambdas.

Unfortunately, there is no such option in the current version of the CDK.

https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-lambda-event-sources.DynamoEventSourceProps.html

When creating manually via the AWS console I see the option below set

```
"functionResponseTypes": [
            "ReportBatchItemFailures"
        ]
```

### Use Case

To enable the creation of lambdas with dynamodb stream as source but leveraging the ReportBatchItemFailure.

https://aws.amazon.com/blogs/compute/optimizing-batch-processing-with-custom-checkpoints-in-aws-lambda/

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

**Comment by mariobittencourt** [comment]
Hi there, I wonder if there is any additional information to help this ticket to move from the triage area to something else.

**Comment by mariobittencourt** [comment]
I checked the CloudFormation documentation and looks like you can use the 'AWS::Lambda::EventSourceMapping' resource for the same. You can enable it by setting the 'FunctionResponseTypes' property to 'ReportBatchItemFailures' in your JSON/YAML template. More details are discussed in the documentation https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-eventsourcemapping.html

**Comment by nija-at** [comment]
Thanks for filing this issue @mariobittencourt.

Yes, this is not currently implemented in the CDK. Until it's available, you can work around this in two different ways.

If you are creating the event source mapping using the `EventSourceMapping` construct, you could do the following -

```ts
const eventSourceMapping = new EventSourceMapping(...); // the dynamo event source mapping you already have


const cfnMapping = eventSourceMapping.node.defaultChild as CfnEventSourceMapping;
cfnMapping.functionResponseTypes = ['ReportBatchItemFailures'];
```

It's a little more tricky if you're using our integration class `DynamoEventSource` - 

```ts
const table = new dynamodb.Table(...); // the table you already have
const fn = new Function(...); // the function you already have
fn.addEventSource(new DynamoEventSource(...)); // the event source you already have


const mapping = fn.node.tryFindChild(`DynamoDBEventSource:${Names.nodeUniqueId(table.node)}`) as EventSourceMapping;
const cfnMapping = eventSourceMapping.node.defaultChild as CfnEventSourceMapping;
cfnMapping.functionResponseTypes = ['ReportBatchItemFailures'];
```

**Comment by mariobittencourt** [comment]
Hi @nija-at thanks for the reply.

I was able to do that using the second snippet of code you provided. Since I have to do the same for a Kinesis stream I am trying to understand, for example, the logic used with the `tryFindChild`.
You used a DynamoDBEventSource string even though the name of the class I used is DynamoEventSource. I assume for KinesisEventSource will be KinesisEventSource but I wonder how to find out for sure.
Regards,
